design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,194 @@
+- a xlang_string with the error message
+- a xlang_string with the error stack if available
+- a xlang_string with an identifier for the projection
+- an IUnknown pointer with any additional language specific information (can be null)
+
+```
+void xlang_originate_error(
+	xlang_result error,
+	xlang_string languageError,
+	xlang_string message,
+	xlang_string stack,
+	xlang_string projectionIdentifier,
+	IUnknown* languageInformation)
+```
+
+If the error is a propagation of a previous xlang error (i.e. the component called another xlang projected component which returned an error), then the associated error info should be retrieved along with the associated xlang error.  The retrieved xlang error should be used as the error instead of translating the language error to a new xlang error.  This will allow the identity of the error to be independent of the translations it may go through as it propagates through various languages and for it to be defined as what the initial projection which returned the error determined it was equivalent to.  Diagnostic information collected by the language should still be added to the error info as it can provide useful information in the case the error becomes a failure.  This can be done by a call to xlang_propagate_error which will add the provided information to the given error info and store it on the thread level storage.
+
+xlang_propagate_error will be defined in the PAL and will take the following parameters:
+- an IXlangErrorInfo* for the error which is being propagated
+- a xlang_string with the language specific error
+- a xlang_string indicating the error stack if available
+- a xlang_string indicating an identifier for the language
+- an IUnknown pointer with any additional language specific information (can be null)
+
+```
+void xlang_propagate_error(
+	IXlangErrorInfo* errorInfo,
+	xlang_string languageError,
+	xlang_string stack,
+	xlang_string projectionIdentifier,
+	IUnknown* languageInformation)
+```
+
+After calling xlang's origination / propagation API, the projection should return the error as the result so that the consuming projection can receive the error and propagate it to the language it projects.  In some cases, the projection may want to call other functions (i.e. cleanup) before returning the error.  If any of these functions error, the projection may treat them as best effort and ignore the error and still propagate the previous error after calling them.  But, those functions may call the error origination API upon error and override the error info stored on the thread local storage.  To avoid this, the projection should retrieve the error info from the thread local storage before doing the best effort work.  Then after doing that work, it should set the retrieved error info back on the thread local store and then return the error.  To help with this, xlang will provide the xlang_get_error_info and xlang_set_error_info API.
```

Comment:
I know we've discussed patterns beyond storing error information in thread local storage. We need to capture those other patterns as well as our reasoning for choosing this approach. DNs are intended to capture why we made design decisions in addition to the actual design decisions themselves.

Reply:
Addressed, I discussed the approaches we considered.